File: python/src/modelos.py
from dataclasses import dataclass


@dataclass
class ConversionRegistro:
    """Datos obtenidos al convertir un número a una base."""

    numero_decimal: int
    base: int
    digitos: list[str]

    def resultado(self) -> str:
        return "".join(self.digitos)


class Conversion:
    """Representa las conversiones de un número decimal."""

    def __init__(self, numero_decimal: int) -> None:
        self.numero_decimal = numero_decimal

    def convertir(self, base: int) -> ConversionRegistro:
        simbolos = "0123456789ABCDEF"

        if self.numero_decimal == 0:
            return ConversionRegistro(0, base, ["0"])

        numero = self.numero_decimal
        residuos: list[str] = []

        while numero > 0:
            residuo = numero % base
            residuos.append(simbolos[residuo])
            numero //= base

        residuos.reverse()
        return ConversionRegistro(self.numero_decimal, base, residuos)

File: typescript/src/modelos.ts
export interface ConversionRegistro {
  numeroDecimal: number;
  base: number;
  digitos: string[];
}

export class Conversion {
  constructor(private readonly numeroDecimal: number) {}

  convertir(base: number): ConversionRegistro {
    const simbolos = "0123456789ABCDEF";

    if (this.numeroDecimal === 0) {
      return { numeroDecimal: 0, base, digitos: ["0"] };
    }

    let numero = this.numeroDecimal;
    const residuos: string[] = [];

    while (numero > 0) {
      const residuo = numero % base;
      residuos.push(simbolos[residuo]);
      numero = Math.floor(numero / base);
    }

    residuos.reverse();
    return { numeroDecimal: this.numeroDecimal, base, digitos: residuos };
  }
}
